File: lab/c/arrays_e_listas/mover_listas.c
#include <stdio.h>

// Função auxiliar para imprimir o vetor
void imprimir(int v[], int n) {
    printf("[");
    for (int i = 0; i < n; i++) {
        printf("%d", v[i]);
        if (i < n - 1) printf(", ");
    }
    printf("]\n");
}


// -------- Versão 1 (usando 2 partes simples) --------
void versao1(int lista[], int n, int k) {
    k = k % n;
    int novo[n];
    int idx = 0;

    // Copia os últimos k
    for (int i = n - k; i < n; i++) {
        novo[idx++] = lista[i];
    }

    // Copia o resto
    for (int i = 0; i < n - k; i++) {
        novo[idx++] = lista[i];
    }

    imprimir(novo, n);
}


// -------- Versão 2 (mesma ideia, mas com vetores separados) --------
void versao2(int lista[], int n, int k) {
    k = k % n;
    int ultimos[k];
    int restantes[n - k];
    int novo[n];
    int i, j;

    // Últimos k elementos
    for (i = 0; i < k; i++) {
        ultimos[i] = lista[n - k + i];
    }

    // Restantes
    for (j = 0; j < n - k; j++) {
        restantes[j] = lista[j];
    }

    // Junta tudo
    for (i = 0; i < k; i++) {
        novo[i] = ultimos[i];
    }
    for (j = 0; j < n - k; j++) {
        novo[k + j] = restantes[j];
    }

    imprimir(novo, n);
}


// -------- Versão 3 (usando índice modular) --------
void versao3(int lista[], int n, int k) {
    k = k % n;
    int novo[n];

    for (int i = 0; i < n; i++) {
        int nova_pos = (i + k) % n;
        novo[nova_pos] = lista[i];
    }

    imprimir(novo, n);
}


// -------- Função principal --------
int main() {
    int lista[] = {1, 2, 3, 4};
    int n = sizeof(lista) / sizeof(lista[0]);

    printf("Versão 1:\n");
    versao1(lista, n, 1);

    printf("\nVersão 2:\n");
    versao2(lista, n, 2);

    printf("\nVersão 3:\n");
    versao3(lista, n, 3);

    return 0;
}

File: lab/javascript/arrays_e_listas/mover_listas.js
// Versão 1 – simples com slice
function versao1(lista, k = 1) {
  k = k % lista.length;
  const ultimos = lista.slice(-k);
  const restantes = lista.slice(0, -k);
  console.log([...ultimos, ...restantes]);
}


// Versão 2 – com loops (sem slice)
function versao2(lista, k = 1) {
  const n = lista.length;
  k = k % n;

  const ultimos = [];
  for (let i = n - k; i < n; i++) {
    ultimos.push(lista[i]);
  }

  const restantes = [];
  for (let j = 0; j < n - k; j++) {
    restantes.push(lista[j]);
  }

  console.log([...ultimos, ...restantes]);
}


// Versão 3 – índice modular
function versao3(lista, k = 1) {
  const n = lista.length;
  const novo = new Array(n);
  k = k % n;

  for (let i = 0; i < n; i++) {
    const novaPos = (i + k) % n;
    novo[novaPos] = lista[i];
  }

  console.log(novo);
}


// Teste
versao1([1, 2, 3, 4], 1);
versao2([1, 2, 3, 4], 2);
versao3([1, 2, 3, 4], 3);
